addons/web/models/ir_actions_server.py

```diff
@@ -0,0 +1,77 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from uuid import uuid4
+from json import JSONDecodeError
+from secrets import compare_digest
+import hashlib
+
+from odoo.http import request
+from odoo import fields, models, api, exceptions, _
+
+
+class ServerAction(models.Model):
+    _inherit = "ir.actions.server"
+
+    url = fields.Char("Public Url", compute="_compute_url")
+    url_path = fields.Char("URL Path", help="Additional URL path", compute="_compute_url_path", store=True)
+    expose = fields.Selection([("webhook", "Expose as webhook")],
+        help="Expose this server action via a controller. The 'as webhook' option will make the server action available via '/web/hook/<url_path>'")
+
+    log_webhook_calls = fields.Boolean(string="Log Calls", default=True)
+
+    @api.constrains("expose", "state")
+    def _constrains_state(self):
+        for action in self:
+            if action.expose == "webhook" and action.state != "code":
+                raise exceptions.ValidationError(_("A webhook is necessarily defined as executing python code"))
+
+    @api.depends("state", "expose", "url_path")
+    def _compute_url(self):
+        for action in self:
+            if action.expose == "webhook" and action.state == "code":
```

Comment:
do we need 'and action.state == "code"' as we have a constraint that enforce webhook and code

Reply:
I suppose we can avoid the state dependency, but it existed in the website code to compute the website_url of the action